### PR TITLE
Initial support for IPv6

### PR DESCRIPTION
**Pull request by macgreagoir**
Adds support for IPv6, but without requiring IPv6.

There are some known issues that should not break, or be broken by, this
commit:

 * Manual is the only supportable provider for IPv6
 * RFC 6761 (6.3), expecting localhost to resolve to both the IPv4 and
   IPv6 loopback addresses, is not always satisfied in Ubuntu builds
   (lp:1644009), but
 * Ubuntu adds both IPv4 and IPv6 loopback addresses regardless of the
   version specified for lo in ENI, so loopback will resolve to a usable
   loopback address in any case
 * Code to be revised when these issues are resolved in Ubuntu are
   commented with `TODO(macgreagoir) IPv6. ...`
 * A bug around `enabled-ha` using existing machines (lp:1642618)
   requires a work-around for HA

QA steps:

 * IPv4, simply bootstrap as usual
 * IPv6
   - Build a machine with a working IPv6 address (which may require something like an IPv6 tunnel from Hurricane Electric) and no IPv4 addresses apart from loopback. (Feel welcome to contact me about this directly.)
   - Deploy to this machine with the manual provider
 * The expected result is a working environment in both IP version cases (noting the HA bug with the manual provider)

### CODE REVIEW

**Comment by macgreagoir** [comment]
!!retry!!

File: apiserver/common/networkingcommon/types.go [hunk 1]
```diff
@@ -479,10 +479,9 @@ func interfaceAddressToNetworkConfig(interfaceName, configType string, address n
 			ipNet = &net.IPNet{IP: ip}
 		}
 	}
-	if ip.To4() == nil {
-		logger.Debugf("skipping observed IPv6 address %q on %q: not fully supported yet", ip, interfaceName)
-		// TODO(dimitern): Treat IPv6 addresses as empty until we can handle
-		// them reliably.
+	if ip.To4() == nil && ip.IsLinkLocalUnicast() {
```
Comment (jameinel) [inline apiserver/common/networkingcommon/types.go]: What do we do with the other IPv6 addresses?
Would it be more obvious if we did:

if ip.To6() != nil && ip.IsLinkLocalUnicast() ?

I almost wonder if for now we'd want to break this up into:

if ip.IP6() != nil {
  if ip.IsLinkLocalUnicast() {
    logger.Debugf("skipping observed IPv6 link-local...") 
  } else {
   logger.Debugf("including ipv6 address %s")
  }
}

File: apiserver/common/networkingcommon/types_test.go [hunk 2]
```diff
@@ -978,6 +978,8 @@ func (s *TypesSuite) TestGetObservedNetworkConfigLoopbackInfrerred(c *gc.C) {
 		ConfigType:    "loopback", // since it is a loopback
 	}, {
 		DeviceIndex:   1,
+		CIDR:          "::1/128",
```
Comment (jameinel) [inline apiserver/common/networkingcommon/types_test.go]: Should this be a new section, rather than just adding to this scenario?

I feel like we should handle IPv4-only loopback as well, shouldn't we?

File: mongo/mongo.go [hunk 3]
```diff
@@ -280,8 +280,16 @@ func SelectPeerAddress(addrs []network.Address) string {
...
+	// TODO(macgreagoir) IPv6. We were always choosing a cloud-local or
+	// localhost addr here (with true arg) but this doesn't make sense in
+	// IPv6, where the IPv6 [public] address is ignored in favour of
+	// ip6-localhost. Only pass true if we find an IPv4 address in the
+	// slice.
 	logger.Debugf("selecting mongo peer hostPort by scope from %+v", hostPorts)
-	return network.SelectMongoHostPortsByScope(hostPorts, true)[0]
+	return network.SelectMongoHostPortsByScope(
+		hostPorts,
+		network.HostPortWithIPv4Address(hostPorts),
```
Comment (jameinel) [inline mongo/mongo.go]: can you use a local variable so that the meaning of this makes more sense? When I read it at a glance it looks like we are only selecting IPv4 addresses.

allowMachineLocal := network.HostPortWithIPv4Address(hostPorts)

return network.SelectMongoHostPortsByScope(hostPorts, allowMachineLocal)

I'm still not 100% sold on using machine local. This feels more like "if we don't *have* a cloud-local then maybe we can fall back to localhost".
Given its the address we are putting into the replicaSet, it seems like it can't ever be right to give a link local address if you are ever actually in a peer group.

File: network/address.go [hunk 4]
```diff
@@ -335,6 +335,20 @@ func SelectMongoHostPortsBySpaces(hostPorts []HostPort, spaces []SpaceName) ([]s
 	return HostPortsToStrings(filteredHostPorts), ok
 }
 
+// HostPortWithIPv4Address returns true if the passed slice of HostPort
+// contains an IPv4 address that is not just a loopback address.
+func HostPortWithIPv4Address(hostPorts []HostPort) bool {
```
Comment (jameinel) [inline network/address.go]: The name sounds like it is returning a HostPort, maybe:

HostPortsHaveIPv4Address()

Not sure what name, but something that makes it clearer the function is returning a boolean.

File: network/address.go [hunk 5]
```diff
@@ -335,6 +335,20 @@ func SelectMongoHostPortsBySpaces(hostPorts []HostPort, spaces []SpaceName) ([]s
...
+	for _, hp := range hostPorts {
+		logger.Debugf("found Address of Value %q, Type %q and Scope %q",
```
Comment (jameinel) [inline network/address.go]: This feels a bit chatty, might be better as Tracef.

File: provider/manual/environ.go [hunk 6]
```diff
@@ -258,7 +258,7 @@ if [ $stopped -ne 1 ]; then
     service %s stop
 fi
 rm -f /etc/init/juju*
-rm -f /etc/systemd/system/juju*
+rm -f /etc/systemd/system{,/multi-user.target.wants}/juju*
```
Comment (jameinel) [inline provider/manual/environ.go]: Does this have anything to do with IPv6 or is it just a drive by that you encountered. Should we be tagging a bug about this so people can know where/when it was fixed?

**Comment by bz2** [comment]
!!retry!!

File: api/apiclient.go [hunk 7]
```diff
@@ -537,11 +540,15 @@ func newWebsocketDialer(cfg *websocket.Config, opts DialOpts) func(<-chan struct
...
-			if !a.HasNext() || isX509Error(err) {
-				// We won't reconnect when there's an X509 error
-				// because we're not going to succeed if we retry
-				// in that case.
-				logger.Infof("error dialing %q: %v", cfg.Location, err)
+			if isX509Error(err) {
+				// We won't reconnect when there's an X509
+				// error because we're not going to succeed if
+				// we retry in that case.
+				logger.Errorf("certificate error dialing %q: %v", cfg.Location, err)
```
Comment (macgreagoir) [inline api/apiclient.go]: It is, aye. It's something I added for clarity in my own testing and thought was worth keeping.

File: mongo/mongo.go (hunk 3 above)
Comment (macgreagoir) [inline mongo/mongo.go]: I'm inclined to agree with you about the original, and continuing IPv4, behaviour. Falling back to loopback/localhost can seem like a bad idea for the replicaset config, though it doesn't cause a problem if HA is never enabled.

Can I suggest that this is machine-local question is a separate bug to be considered outside this change, as a new behaviour?

Comment (macgreagoir) [inline mongo/mongo.go]: Local variable done.

File: network/address.go (hunk 4 above)
Comment (macgreagoir) [inline network/address.go]: Done.

File: network/address.go (hunk 5 above)
Comment (macgreagoir) [inline network/address.go]: Done.

File: provider/manual/environ.go (hunk 6 above)
Comment (macgreagoir) [inline provider/manual/environ.go]: Fair call :-) I really don't like drive-by changes myself.

PR juju/juju/pull/6624

File: apiserver/common/networkingcommon/types_test.go (hunk 2 above)
Comment (macgreagoir) [inline apiserver/common/networkingcommon/types_test.go]: I'm not sure this is still a likely use case. I see both versions everywhere.

**Comment by macgreagoir** [comment]
!!retry!!

**Comment by bz2** [comment]
!!retry!!

**Comment by macgreagoir** [comment]
$$merge$$

**Comment by macgreagoir** [comment]
@bz2 Thanks for the manual stop :-) I'll $$merge$$ with squash now.
